job-server/src/main/scala/spark/jobserver/JobManagerActor.scala

```diff
@@ -496,4 +570,89 @@ class JobManagerActor(daoActor: ActorRef, supervisorActorAddress: String, contex
+   *    contains this option, is used here.
+   */
+  private def restartTerminatedJobs(contextId: String, senderRef: ActorRef): Unit = {
+    (daoActor ? JobDAOActor.GetJobInfosByContextId(
+        contextId, Some(Seq(JobStatus.Running, JobStatus.Error))))(daoAskTimeout).onComplete {
+      case Success(JobDAOActor.JobInfos(Seq())) =>
+        logger.info(s"No job found for context ${contextName} which was terminated unexpectedly." +
+            " Not restarting any job.")
+      case Success(JobDAOActor.JobInfos(jobInfos)) =>
+        logger.info(s"Found jobs for this context ${contextId}")
+
+        val restartCandidates = getRestartCandidates(jobInfos)
+        logger.info(s"Total restart candidates are ${restartCandidates.length}")
+        totalJobsToRestart = restartCandidates.length
+        restartCandidates.foreach { jobInfo =>
+          (daoActor ? JobDAOActor.GetJobConfig(jobInfo.jobId))(daoAskTimeout).onComplete {
+            case Success(JobDAOActor.JobConfig(Some(config))) =>
+              restartJob(jobInfo, config)
+            case Success(JobDAOActor.JobConfig(None)) =>
+              updateJobInfoAndReportFailure(jobInfo, NoJobConfigFoundException(jobInfo.jobId))
+            case Failure(e: Exception) =>
+              // In case of error during job restart, an error will be written to the DAO.
+              // In case that this job is a dependency for another job, it may cause problems.
+              // A strategy must be implemented at this point.
+              updateJobInfoAndReportFailure(jobInfo, e)
+            case _ =>
+              updateJobInfoAndReportFailure(jobInfo, UnexpectedMessageReceivedException(jobInfo.jobId))
+          }
+        }
+      case Failure(e: Exception) =>
+        logger.error(s"Exception occured while accessing job ids for context ${contextId} from DAO.", e)
+        self ! PoisonPill
+      case unexpectedMsg @ _ =>
+        logger.error(s"Unexpected scenario occured, message received is $unexpectedMsg")
+        self ! PoisonPill
+    }
+  }
+
+  private def updateJobInfoAndReportFailure(jobInfo: JobInfo, error: Exception) {
+    updateJobInfoWithErrorState(jobInfo, error)
+    self ! JobRestartFailed(jobInfo.jobId, error)
+  }
+
+  private def updateJobInfoWithErrorState(jobInfo: JobInfo, error: Throwable) {
+   val updatedJobInfo = jobInfo.copy(state = JobStatus.Error,
+     endTime = Some(DateTime.now()), error = Some(ErrorData(error)))
+   daoActor ! JobDAOActor.SaveJobInfo(updatedJobInfo)
+  }
+
+  private def restartJob(existingJobInfo: JobInfo, existingJobConfig: Config) {
+    // Add response to master for testing
+    var respMsg = s"Restarting the last job (JobId=${existingJobInfo.jobId} & "
+    respMsg = respMsg + s"contextName=${existingJobInfo.contextName})"
+    logger.info(respMsg)
+    val fullJobConfig = existingJobConfig.withFallback(config).resolve()
+    val contextConfig = Try(existingJobConfig.getConfig("spark.context-settings"))
```

Comment:
It seems to me a good idea to have code/methods to get context settings out.  More typesafe than having strings everywhere.

Reply:
Good point. I will try to adapt this change.